### ISSUE TITLE
imbalanced dataset

### ISSUE THREAD

**Issue by SoodabehGhaffari** [chemprop/chemprop#416]
Hello,
How can we use Chemprop for an imbalanced dataset? Do we need to do anything specific? Is scaffold splitting suitable for imbalanced dataset?
Also, can we use ensembling for predefined data split?

Thank you

**Comment by kevingreenman** [comment]
Chemprop doesn't have any built-in options for handling imbalanced data. You would need to do any up-sampling, down-sampling, or other methods as preprocessing steps before giving the data to chemprop. 

It's hard to say whether scaffold splitting would be suitable without knowing more about the distribution of labels with respect to scaffolds in your particular dataset (e.g. do all the over-represented samples come from the same scaffold?).

Yes, you can use ensembling with any type of data split. Chemprop ensembles work by using different random initialization of the network weights for each member of the ensemble, and the same data splits will be used for each.

**Comment by kevingreenman** [comment]
Correcting my previous error I discovered while responding to #572...there actually is a (not very well documented) option in chemprop for handling imbalanced classification data: the `--class_balance` flag. However, this is the only option and doesn't give you much flexibility, which is why using external splitting might be more desirable in this case.

**Comment by SoodabehGhaffari** [comment]
Thank you, Kevin.
Just one more question:
Do we need to set --class_balance *True* to address the imbalanced dataset?

Best Regards
Soodabeh

**Comment by kevingreenman** [comment]
`--class_balance` is [defined](https://github.com/chemprop/chemprop/blob/9bc0d0ef483bd6e43ab097bbb5b93a7b065f1fa2/chemprop/args.py#L465) as a `bool`, so including the flag itself in your command is enough to tell Chemprop you want it to be `True` (i.e., you should not include the word `True` after the flag).

**Comment by SoodabehGhaffari** [comment]
Thank you for the help. I appreciate it.
I already split the data based on the same positive class in my training,
validation, and test and optimized hyperparameters based on the combination
of my training and validation data.

Here is the command I am using:

chemprop_train --data_path
/scratch/gpfs/sg6615/chemprop/data/df_binary_train.csv --separate_val_path
/scratch/gpfs/sg6615/chemprop/data/df_binary_val.csv --separate_test_path
/scratch/gpfs/sg6615/chemprop/data/df_binary_test.csv --dataset_type
classification --save_dir
/scratch/gpfs/sg6615/chemprop/classification/20_ensemble/train_checkpoint
--config_path /scratch/gpfs/sg6615/chemprop/classification/my-config.json
--smiles_column SMILES_FROM_INCHI --target_columns target
--features_generator rdkit_2d_normalized --no_features_scaling
--ensemble_size 20 --class_balance




Is there any other step/flag that I need to add? Do I need to do any cross
validation? Please let me know if I need to make any changes.




Best Regards

Soodabeh

**Comment by SoodabehGhaffari** [comment]
Thank you, Kevin, for the help. I appreciate it.
I already split the data based on the same positive class in my training,
validation, and test and optimized hyperparameters based on the combination
of my training and validation data.

Here is the command I am using:

chemprop_train --data_path
 /scratch/gpfs/sg6615/chemprop/data/df_binary_train.csv --separate_val_path
 /scratch/gpfs/sg6615/chemprop/data/df_binary_val.csv --separate_test_path
 /scratch/gpfs/sg6615/chemprop/data/df_binary_test.csv --dataset_type
 classification --save_dir
 /scratch/gpfs/sg6615/chemprop/classification/20_ensemble/train_checkpoint
--config_path /scratch/gpfs/sg6615/chemprop/classification/my-config.json
--smiles_column SMILES_FROM_INCHI --target_columns target
--features_generator rdkit_2d_normalized --no_features_scaling
--ensemble_size 20 --class_balance




Is there any other step/flag that I need to add? Do I need to do any cross
validation? Please let me know if I need to make any changes.




Best Regards

Soodabeh

**Comment by spyda90** [comment]
Hi Soodabeh,

I am the author of the question in https://github.com/chemprop/chemprop/issues/572.

If I understood correctly you are just splitting the "minority" class into training, val, test sets to be sure it is quite well represneted in all of them, without performing any oversampling / undersampling process, right?

If that is the case I don't know how much you can improve the performance of your model. But of course it is mainly dependent on how much the dataset is imbalanced.

Hope this helps.

Francesco

**Comment by SoodabehGhaffari** [comment]
Hi Francesco,
Thank you for the response.
That is right. I did not oversampling and under sampling. How I split the
positive class equally in training, validation and test.
My data is extremely imbalanced ( one percent positive).
Do I need to do oversampling/undersampling? If so, Could you give me a
suggestion how to do that?

Best Regards
Soodabeh

**Comment by kevingreenman** [comment]
@SoodabehGhaffari I agree with @spyda90, just splitting the "minority" class into training, val, and test sets to be sure it is well represented in all of them (without any oversampling / undersampling) will likely not help you. I suggest you look at the resources I linked in #572 and the blog post @spyda linked in #572 for suggestions on how to do oversampling/undersampling.

Other than that, your training command looks okay. Whether or not you do cross validation is up to you. It is generally good to do and gives you a more robust result.

**Comment by SoodabehGhaffari** [comment]
Hi kevin,
I am reaching out to ask for help with cross-validation for an imbalanced
dataset. We discussed it before, however, I still need help and
appreciate your help. I split my dataset into training, validation, and
test data with the same percentage of positive class. Then, I oversampled
my minority class using RandomOverSampler in only the training data such
that the ratio minority class to the majority in the training is now 5% (in
the original data it was below 1%). To add the cross-validation, is it
correct to add the cross validation to the code below as (--num_folds 10)?
Without cross validation and just oversampling, my model improved.

 In the post, you suggested making 10 separate jobs for each fold and
taking an average of the metrics. However, in that way, we get 10 separate
ensemble models (assuming we use an ensemble in our modeling). I do not
know how to proceed with 10 separate ensemble models. Should we apply them
to our new dataset separately and take an average for the output?

I really appreciate your help.

Best Regards
Soodabeh

chemprop_train --data_path ./df_train_antibiotics_over_5%.csv
--separate_val_path
/scratch/gpfs/sg6615/chemprop/data/df_binary_antibiotics_val.csv
--separate_test_path
/scratch/gpfs/sg6615/chemprop/data/df_binary_antibiotics_test.csv
--dataset_type classification --save_dir
/scratch/gpfs/sg6615/chemprop/oversample/antibiotics_train_checkpoint_5%
--config_path
/scratch/gpfs/sg6615/chemprop/classification_antibiotics/antibiotic-config.json
--smiles_column SMILES_FROM_INCHI --target_columns binary_Antibiotics
--features_generator rdk it_2d_normalized --no_features_scaling
--ensemble_size 20 --class_balance --metric prc-auc
